File: ListaExercício1/Python/Ex2.py
import random
random.seed(123)

def calculaCustoDano(tipo, vida):
    if tipo == "Leve":
        return (400, vida-0.01)
    elif tipo == "Moderado":
        return (500, vida-0.03)
    else:
        return (700, vida-0.07)
    

def simulacao():
    lTipos = ["Leve", "Moderado", "Severo"]
    vidaUtil = 1.0
    d_controle = {'Leve': [0,0], 'Moderado': [0,0], 'Severo': [0,0], 'Total': [0,0]}
    for _ in range(720):

        falhaTotal = random.random()
        if falhaTotal<=0.002:
            vidaUtil = 0
            tipoDano = "Total"
            custo = 2000

        else:
            tipoDano = random.choices(lTipos, weights=(0.7,0.2,0.1))[0]
            custo, vidaUtil = calculaCustoDano(tipoDano, vidaUtil)
    

        if vidaUtil<=0:
            d_controle[tipoDano][0] += custo
            d_controle[tipoDano][1] += 1 #contador
            vidaUtil = 1

    return d_controle

media_substituicao = 0
media_custo = 0
media_falhas = 0
for _ in range(1000):
    d = simulacao()

    media_substituicao += (d.get('Leve')[1]+d.get('Moderado')[1]+d.get('Severo')[1]+d.get('Total')[1])
    media_custo += (d.get('Leve')[0]+d.get('Moderado')[0]+d.get('Severo')[0]+d.get('Total')[0])
    media_falhas += d.get('Total')[1]

print("A média de substituições por simulação: ", media_substituicao/1000)   
print("O custo médio total por simulação: ", media_custo/1000)
print("O número médio de falhas totais aleatórias: ", media_falhas/1000)           


# -------------------------

# import random

# #cada chamada a essa função significa uma execução de 720h
# def simulacao():
#     random.seed(123)
#     vidautil = 1.0
#     custo = 0
#     contSubstituicao = 0
#     custoSub=0
#     falha=0
#     for i in range(720):
#         probabilidade = random.randint(0, 100)
#         crash = random.randint(0, 1000)

#         if 0<=crash<=2:
#             vidautil = 1
#             custo+=2000
#             falha +=1
#             contSubstituicao+=1

#         if vidautil<=0:
#             vidautil = 1
#             custo += custoSub
#             contSubstituicao+=1

#         else:
#             if probabilidade<=10:
#                 vidautil-=0.07
#                 custoSub = 700

#             if 10<probabilidade<=30: 
#                 vidautil-=0.03
#                 custoSub = 500

#             if 30<probabilidade<=100:
#                 vidautil-=0.01
#                 custoSub = 400

#     return falha, custo, contSubstituicao

# mediaSim=0
# mediaCusto=0
# totalFalhas=0
# for i in range(1000):
#     falha, custo, cont = simulacao()
#     totalFalhas += falha
#     mediaSim += cont
#     mediaCusto += custo/cont

# print("Media de substituições por simulação: ",mediaSim/1000)
# print("Custo medio por simulação: ",mediaCusto/1000)
# print("Quantidade total de falhas aleatorias: ",totalFalhas/1000)

File: ListaExercício1/R/Ex2.r
set.seed(123)

cont_falhas_totais <- 0
substituicoes <- 0
custo_total <- 0

for (i in 1:1000){
    vida_util <- 1.0
    for (h in 1:720){
        prob <- runif(1)
        if (prob <= 0.002){
            cont_falhas_totais <- cont_falhas_totais + 1
            substituicoes  <- substituicoes  + 1
            custo_total <- custo_total + 2000
            vida_util <- 1.0
        }
        else{
            red_vida <- sample(c(0.01,0.03,0.07),size=1,prob=c(0.70,0.20,0.10))
            vida_util <- vida_util-red_vida
            if (vida_util<=0){
                substituicoes  <- substituicoes  + 1
                if (red_vida == 0.01){
                    custo = 400
                }
                if (red_vida == 0.03){
                    custo = 500
                }
                if (red_vida == 0.07){
                    custo = 700
                }
                custo_total <- custo_total + custo 
                vida_util <- 1.0
            }
        }
    }
}

cat("Media de substituição por simulação: ", substituicoes /1000)
cat("O custo médio total por simulação: ", custo_total/1000)
cat("O número médio de falhas totais aleatórias: ", cont_falhas_totais/1000)
cat("\n")